design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+operation will not be available in headless or embedded mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. 
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+If the user specifies `AVRO` encoding in their `CREATE` statement, we will register the topic with 
+the schema registry proactively if the topic did not previously exist. Note that this will not 
+change how we handle `CREATE` statements for topics that already exist (today we do not validate
+that the schema supplied matches the schema in Schema Registry).
+
+### ROWKEY and ROWTIME Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT VALUES`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "timestamp": 1, "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. 
+
+The column schema will always include the implicit `ROWKEY` and `ROWTIME` fields, and the user can
+choose to specify or ignore these fields. To prevent the inconsistencies described above, though,
+we will automatically enforce consistency whenever the key property is set in the `WITH` clause.
+
+* If the user sets a `KEY` property, they do not need to supply `ROWKEY` in their `INSERT INTO`
+statement (e.g. `INSERT INTO foo (id, foo) VALUES ("key", 123);`). In this case, the kafka message
+produced will automatically populate `ROWKEY` to the value of `id`: 
+  ```json
+  {"key": "key", "timestamp": 1, "value": {"id": "key", "foo": 123}}
+  ```
+* Analogous to the above, if the user provides the key but not `ROWKEY`, it will be populated for 
+them.
+* If the user sets a `KEY` property _and_ they supply `ROWKEY` in their `INSERT INTO` statement, 
+KSQL will enforce that the value of `ROWKEY` and the value of the key column are identical.
+  ```sql
+  INSERT INTO foo (ROWKEY, ID, VALUE) VALUES ("THIS WILL FAIL", "key", 123);
+  ```
+* If the user does not set a `KEY` property, and they are inserting into a stream, they may choose
+to omit `ROWKEY` altogether. This will result in a `null` key.
+* If the user does not set a `KEY` property, and they are inserting into a table, they _must_
```

Comment:
I think we should avoid differences between streams and tables where ever possible. So my question would be... do we allow a `null` key in a table?  If we do, then we should allow `null` keys when inserting into a table, so we could just do away with this inconsistency.

Even if the code doesn't allow `null` keys in a tables at present, we may in the future... `NOT NULL` etc..

What's more - once structured keys comes along, it would be possible for some of the key fields to be supplied, but not others, which again makes the special handling here questionable.

Reply:
Okay - I'm happy not adding this extra restriction. Based on Rohan's comment on another PR looks like there is a case with `null` keys in tables.